include/odp/api/spec/comp.h

```diff
@@ -0,0 +1,661 @@
+	 *
+	 * Application selects which mode will be used for
+	 * compression/decompression operation.
+	 *
+	 * @see odp_comp_op(), odp_comp_op_enq()
+	 */
+	odp_comp_op_mode_t mode;
+
+	/** Compression algorithm
+	 *
+	 *  @see odp_comp_capability()
+	 */
+	odp_comp_alg_t comp_algo;
+
+	/** Hash algorithm
+	 *
+	 *  @see odp_comp_alg_capability()
+	 */
+	odp_comp_hash_alg_t hash_algo;
+
+	/** parameters specific to compression */
+	odp_comp_alg_param_t algo_param;
+
+	/** Session packet enqueue ordering
+	 * Boolean to indicate if packet enqueue ordering is required per session
+	 * If packet_order is false, packet order is maintained on
+	 * operation completion order and not in session enqueue order.
+	 * By disabling packet order requirement, performance oriented
+	 * application can leverage HW offered parallelism to increase operation
+	 * performance.
+	 */
+	odp_bool_t packet_order;
+
+	/** Destination queue for de/compression completion result
+	 *
+	 * Operations in asynchronous mode enqueue resulting events into this
+	 * queue. Results are enqueued as ODP_EVENT_PACKT with subtype
+	 * ODP_EVENT_PACKET_COMP
+	 */
+	odp_queue_t compl_queue;
+} odp_comp_session_param_t;
+
+/**
+ * Comp packet operation result
+ */
+typedef struct odp_comp_packet_result_t {
+	/** Operation Return Code */
+	odp_comp_status_t status;
+
+	/** Output packet data range */
+	odp_packet_data_range_t out_data_range;
+} odp_comp_packet_result_t;
+
+/**
+ * Comp API per packet operation parameters
+ */
+typedef struct odp_comp_packet_op_param_t {
+	/** Session handle from creation */
+	odp_comp_session_t session;
+
+	/** Boolean indicating if current input is last chunk of data, where
+	 *
+	 *   true : last chunk
+	 *
+	 *   false: more to follow
+	 *
+	 * If set to true, indicates this is the last chunk of
+	 * data. After processing of last chunk of data is complete i.e.
+	 * call returned with any error code except ODP_COMP_ERR_OUT_OF_SPACE,
+	 * implementation should move algorithm to stateless mode
+	 * for next of batch of operation i.e. reset history,
+	 * insert 'End of Block' marker into compressed data stream(if
+	 * supported by algo).See deflate/zlib for interpretation of
+	 * stateless/stateful.
+	 *
+	 * For stateless compressions (ex ipcomp), last should be set to 'true'
+	 * for every input packet processing call.
+	 *
+	 * For compression + hash, digest will be available after
+	 * last chunk is processed completely. In case of
+	 * ODP_COMP_ERR_OUT_OF_SPACE, application should keep on calling
+	 * odp_comp_xxx() API with more output buffer unless call returns
+	 * with ODP_COMP_ERR_NONE or other failure code except
+	 *  ODP_COMP_ERR_OUT_OF_SPACE.
+	 */
+	odp_bool_t last;
+
+	/** Input data range to process */
+	odp_packet_data_range_t in_data_range;
+
+	/** Output packet data range.
+	 * Indicates where processed packet will be written
+	 * where length specifies, data available for output.
+	 */
+	odp_packet_data_range_t out_data_range;
```

Comment:
Unclear how this is expected to be used. Suppose at input `odp_packet_len(pkt)` = 100 and a compression operation specifies an `in_data_range` of 20..39. Since compression (should) result in fewer output bytes than input bytes, what should be used as the `out_data_range`? If this is specified as 20..39 (same as the input), what's going to be the `odp_packet_len()` of the output packet? If those 20 input bytes compress down to 10 bytes, what does the resulting output packet look like? Presumably I'd want the packet to consist of the original bytes 0..19, then bytes 20..29, which represent the compressed `in_data_range`, and then the original bytes 40..99 which now occupy offsets 30..89 in the output packet which now has an `odp_packet_len()` of 90. But is this what happens?

Can the `out_data_range` be disjoint from the `in_data_range`? Can it partially overlap? And exactly what happens in these cases in the output packet?

Similar questions arise for decompression. If I say decompress `in_data_range` 20..39 how is `out_data_range` specified such that these bytes are decompressed and the following packet bytes remain unchanged?

I think we need some practical examples here to illustrate exactly how these parameters should be used to achieve a desired result.

Reply:
The above comments seem equally applicable to this version of the PR. It's not at all clear how these APIs can be used to compress or decompress partial packets because it seems its undefined whether bytes not in the specified range will be altered as a result of these operations.

Should we have a simplified set of initial APIs that compress or decompress entire input packets?

Reply:
Agree, output range is useful only if bytes outside of the range must not be modified during the operation.

Reply:
The compression engine will not access any other data in the input and output packet apart from the data_range given as input. Since the compression engine works on buffers and files the engine does not understand the concept of a packet.
output_data_range is used to specify the location on the output buffer where the compressed/decompressed output will be written. The compression engine only checks the input_data_range in input packet and compresses the same and writes the output to output packet in the output_data_range. The rest of the bytes of the input packet are not copied to output packet since the format could be completely different from input packet. Neither does the compression engine update the packet length, headroom or tailroom.

'output_data_range' is required since the entire output packet is not accessed by the compression engine and it will only update the result on output packet within the range provided. i.e application could provide output_data_range from offset = 100 and length = 200, for one input packet and can use the same output packet with data range offset = 300 and length = 200 for another input packet.

Reply:
Text above needs to specify that "output packet data is not modified outside of the provided out_data_range"

Reply:
I had this documentation in a previous version and somehow lost during the update. Will update in next version